### PR TITLE
CICADA-uGT emulator additions

### PR DESCRIPTION
**Pull request by aloeliger**
### PR description:

This PR adds the uGT half of the CICADA -> uGT emulator chain. It adds all the pieces needed to parse and compare the new CICADA condition added in uTM v 0.12.0. This is being opened in draft because no unpacker/packer combo is available yet, but will be added after I have had the chance to speak to experts about reading the CICADA record.

Depends on utm version 0.12.0

@elfontan You should likely take a look at my work on this, this is more working with uGT than I am used to.

@cms-sw/hlt-l2 Just in case you aren't automatically pinged on this, FYI. There is no unpacker/packer, but I will keep this in draft while I develop them. I will be speak to experts about it ASAP.

EDIT 06/03/24: Draft unpacker (not non-trivially tested) added.

### PR validation:

All code compiles and has had code formatting applied.

In addition, I have tested the emulator on a new proto-type menu including a CICADA seed:
https://cernbox.cern.ch/files/link/public/LeBbgndwFDjWoI5?tiles-size=1&items-per-page=100&view-mode=resource-table

The emulator runs all the way through. The following are debug statements (now removed) from various components of the global board, CICADA conditions, and trigger menu parser showing that the emulator receives the score, compares to the menu, and in this case, decides it is below threshold and doesn't pass the menu cut:

```
Begin processing the 2nd record. Run 368566, Event 2615907242, LumiSection 1195 on stream 0 at 27-Feb-2024 08:34:54.517 CST
GlobalBoard::receiveCaloObjectData receive cicada
Valid CICADA
2.73438
GlobalBoard::runGTL CondCICADA
CICADACondition::evaluationCondition begin
cicada score is: 2.73438
min threshold is: 42
passCondition? 0
```

### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This PR is not a backport however, is likely to be needed for data-taking so a backport is probable to CMSSW_14_0.

### CODE REVIEW

**Comment by aloeliger** [comment]
I'm marking this pull request ready after adding the unpacker commit. However, I should stress that the unpacker has not been tested in realistic conditions with actual CICADA input data present, and therefore has not been tested to be working. It does compile, and it does seem to properly fail-safe to a 0.0 score in the case of nothing to unpack/nothing unpacked.

**Comment by aloeliger** [comment]
assign hlt

- I think HLT wants to sign off on changes to the uGT emulator?

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  //And we shift by 4 bits a time, hence the factor of 4
+  for (unsigned short bitIndex = 0; bitIndex < numCICADAWords; ++bitIndex) {
+    constructedScore += ((float)bitsArray[bitIndex]) * pow(2.0, 4 * (numCICADAWords / 2 - (bitIndex + 1)));
+  }
+  return constructedScore;
+}
+
+bool l1t::stage2::CaloSummaryUnpacker::unpack(const Block& block, UnpackerCollections* coll) {
+  LogDebug("L1T") << "Block ID  = " << block.header().getID() << " size = " << block.header().getSize();
+  std::cout << "Unpacking CICADA!" << std::endl;
```
Comment (Martin-Grunewald) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:29]: cout can not be in production code - use LogDebug or so.

Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:29]: Ahh, my bad. I know better than that. Forgot the debug statement. Thanks for catching that.

**Comment by Martin-Grunewald** [comment]
It is not clear to me how/where the unpacker is called. Within existing (unpacking) code or a new module, which needs to be added to some python unpacking sequence? What about the packer?

**Comment by aloeliger** [comment]
> It is not clear to me how/where the unpacker is called. Within existing (unpacking) code or a new module, which needs to be added to some python unpacking sequence? What about the packer?

@Martin-Grunewald It's defined as part of the GT unpacking setup here: https://github.com/cms-sw/cmssw/blob/c0b5b266322de4b536b3e8bd7d41749da8b74764/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc#L131

I need to wait until I have test crate records before I can test this non-trivially however. (Seems like it wont unpack empty/unused records?).

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h [hunk 2]
```diff
@@ -40,6 +43,7 @@ namespace l1t {
       inline EtSumBxCollection* getZDCSums(const unsigned int copy) override { return zdcsums_[copy].get(); };
       inline JetBxCollection* getJets(const unsigned int copy) override { return jets_[copy].get(); };
       inline TauBxCollection* getTaus(const unsigned int copy) override { return taus_[copy].get(); };
+      inline float* getCICADAScore() override { return cicadaScore_.get(); };
```
Comment (Martin-Grunewald) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:46]: Generl question associated to the above: Is it clear that this will ever be 'just' a float, not in future calculating several values for different anomaly calo detectors, and/or for different BXs?

Comment (Martin-Grunewald) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:46]: ```float*``` seems overkill, ```float``` would do? But ok, this pattern is used for the other products.

Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:46]: > General question associated to the above line: Is it clear that this will ever be 'just' a float, not in future calculating several values for different anomaly calo detectors/algorithms, and/or for different BXs?

It's a good question actually. From the Calo summary side (the card being plugged into uGT), we don't make BX distinctions, however it is/was unclear to me whether the typical BX behavior of L1T was a product of sub-systems feeding that information to uGT, or uGT retaining BX information and the emulators shuffling that responsibility to someone else for convenience. In the first case, and it is my understanding that this is what happens, we don't need to unpack other CICADA/summary BXs, but if it were the second, we would. Again, to the best of my ability to figure out, I think we do not need to construct BX related info, or expect to see it in the uGT input record, because CICADA will not provide it.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc (hunk 1 above)
Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc]: Fixed in recent (squashed) commit

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by Martin-Grunewald** [comment]
code-checks

**Comment by aloeliger** [comment]
CICADA is likely coming online for testing in the near future. I have rebased this PR to fix the conflicts.

uGT has also provided a more accurate input record example for what CICADA needs to unpack. I will be making the necessary changes ASAP and testing as soon as I am able.

**Comment by aloeliger** [comment]
Okay. After discussion with uGT experts, and a quick reference on what the input record for the calosummary/CICADA card is expected to look like (included below), I have updated the unpacker and CICADA emulator to take advantage of all BXs by changing emulator outputs and unpacker outputs to BX vectors (the uGT changes have been updated to use this BX vector & grab the only guaranteed available BX, the central one). The emulator will not, and really cannot create information for BXs outside the central one, but the unpacked CICADA BXs may be useful for other studies around CICADA (pre-firing comes to mind?) At this point, this unpacker has still not been thoroughly tested (I am searching for a way to verify it ASAP), but it should now theoretically be complete.

Reference on CICADA input records at uGT:
```
161e0200xxxxxxxx => "16" indicates begin of CICADA input data
2000000010000000 => AD I LSB (cicada frame 1, bx-2) & AD I MSB (cicada frame 0, bx-2)
4000000030000000 => AD D LSB (cicada frame 3, bx-2) & AD D MSB (cicada frame 2, bx-2)
0000000000000000 => free (cicada frame 5, bx-2) & free (cicada frame 4, bx-2)
6000000050000000 => AD I LSB (cicada frame 1, bx-1) & AD I MSB (cicada frame 0, bx-1)
8000000070000000 => AD D LSB (cicada frame 3, bx-1) & AD D MSB (cicada frame 2, bx-1)
0000000000000000 => free (cicada frame 5, bx-1) & free (cicada frame 4, bx-1)
a000000090000000 => AD I LSB (cicada frame 1, bx) & AD I MSB (cicada frame 0, bx)
c0000000b0000000 => AD D LSB (cicada frame 3, bx) & AD D MSB (cicada frame 2, bx)
0000000000000000 => free (cicada frame 5, bx) & free (cicada frame 4, bx)
e0000000d0000000 => AD I LSB (cicada frame 1, bx+1) & AD I MSB (cicada frame 0, bx+1)
f0000000f0000000 => AD D LSB (cicada frame 3, bx+1) & AD D MSB (cicada frame 2, bx+1)
0000000000000000 => free (cicada frame 5, bx+1) & free (cicada frame 4, bx+1)
8000000070000000 => AD I LSB (cicada frame 1, bx+2) & AD I MSB (cicada frame 0, bx+2)
1000000020000000 => AD D LSB (cicada frame 3, bx+2) & AD D MSB (cicada frame 2, bx+2)
0000000000000000 => free (cicada frame 5, bx+2) & free (cicada frame 4, bx+2)
xxxxxxxx181e0100 => "18" indicates begin of EXTERNAL CONDITION input data

"AD I": means Anomaly Detection Integer part (8 bits).
"AD D": means Anomaly Detection Decimal part (8 bits).
"MSB": means the 4 most significant bits of the 8 bits word (bits 31..28 in record), 
"LSB": means the 4 least significant bits of the 8 bits word (bits 63..60 in record).
(Bits 59..32 and 27..0 of frames 0..3 and all bits of frames 4..5 are always zero).
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h (hunk 2 above)
Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h]: > General question associated to the above line: Is it clear that this will ever be 'just' a float, not in future calculating several values for different anomaly calo detectors/algorithms, and/or for different BXs?

After discussion with uGT, it is not "just a float", and modifications to the setup have been made from 49ad545fc7b871b2e01e084431ff04f0778b70c7 on. See https://github.com/cms-sw/cmssw/pull/44222#issuecomment-2054144866

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
This PR may need to be split into an emulator and unpacker portion so the emulator can allow the creation of test crate menu firmware (https://its.cern.ch/jira/browse/CMSLITDPG-1221), and the test crate firmware can be used to verify the unpacker. The emulator will need to be backported and integrated first before any testing can proceed.

**Comment by mmusich** [comment]
@aloeliger what's the status of this and / or of the companion https://github.com/cms-sw/cmssw/pull/45037 ?

**Comment by aloeliger** [comment]
@mmusich We're still waiting for proper alignment and timing studies of the summary CTP7s to the uGT cards and DAQ path, which is preventing CICADA commissioning. Until that happens I can't verify the unpacker/emulator chain against hardware values bit by bit. Strictly, it's ready, and the algorithm here "should" work.

**Comment by aloeliger** [comment]
Okay, the unpacker has now been tested on real uGT data, and seems to work
image
This is a FED dump of a FED 1405 event, with the CICADA data highlighted

image
This is debug output from the unpacker, showing that it gets the proper words, in the proper order for the right BX, and the conversion from hex bits to a decimal number seems to be behaving properly according to a hex to decimal converter.

The dumping of the collection into the event cannot be tested until we actually plug into uGT, which requires this beforehand, so all possible tests have been exhausted, and the unpacker seems to work.

**Comment by aloeliger** [comment]
+l1

**Comment by aloeliger** [comment]
@cms-sw/hlt-l2 Just an FYI, this is theoretically ready as of now, and (with my apologies) it is also I think on a razor thin timescale. I will be preparing a backport as soon as I am able. Sorry again about that.

File: L1Trigger/L1TGlobal/src/GlobalBoard.cc [hunk 3]
```diff
@@ -342,6 +346,21 @@ void l1t::GlobalBoard::receiveCaloObjectData(const edm::Event& iEvent,
...
   }
+  if (receiveCICADA) {
+    edm::Handle<BXVector<float>> cicadaScoreHandle;
+    iEvent.getByToken(CICADAInputToken, cicadaScoreHandle);
+    if (not cicadaScoreHandle.isValid()) {
+      if (m_verbosity) {
+        edm::LogWarning("L1Tglobal") << "\nWarning: Input tag for the CICADA score"
+                                     << "\nrequested in configuration, but not found in the event.\n"
+                                     << "\nSetting score to 0.0";
+        setCICADAScore(0.0);
```
Comment (Martin-Grunewald) [inline L1Trigger/L1TGlobal/src/GlobalBoard.cc:357]: This setting to zero should be outside of the if block printingthe message.

Comment (aloeliger) [reply L1Trigger/L1TGlobal/src/GlobalBoard.cc:357]: Thank you, good catch.

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
+l1

File: L1Trigger/L1TGlobal/src/GlobalBoard.cc [hunk 4]
```diff
@@ -342,6 +346,21 @@ void l1t::GlobalBoard::receiveCaloObjectData(const edm::Event& iEvent,
...
+      if (m_verbosity) {
+        edm::LogWarning("L1Tglobal") << "\nWarning: Input tag for the CICADA score"
+                                     << "\nrequested in configuration, but not found in the event.\n"
+                                     << "\nSetting score to 0.0";
+      }
+      setCICADAScore(0.0);
+    } else
+      setCICADAScore(cicadaScoreHandle->at(
+          0,
+          0));  //CICADA emulation will only provide a central BX, and one value. Unpacking may have more values, but that can't be guaranteed.
```
Comment (missirol) [inline L1Trigger/L1TGlobal/src/GlobalBoard.cc:362]: Is it guaranteed that this will never crash ? In other words, should a check like
```cpp
if (not cicadaScoreHandle->isEmpty(0))
```
be added ?

Comment (aloeliger) [reply L1Trigger/L1TGlobal/src/GlobalBoard.cc:362]: The only reason this would occur is there is a menu with CICADA in it, but there is no data in the uGT card, which is unlikely in a production scenario.

That being said, I did chase my tail for about an hour last night because I forgot this fact while trying to validate the unpacker collections before I realized there was no production unpacker stuff to validate.

I think it's better safe than sorry, so I can/will make this addition.

Comment (aloeliger) [reply L1Trigger/L1TGlobal/src/GlobalBoard.cc:369]: Should be a guard for this in the most recent commit

**Comment by aloeliger** [comment]
please test

File: DataFormats/L1CaloTrigger/src/classes_def.xml [hunk 5]
```diff
@@ -13,4 +13,8 @@
   <class name="std::vector<L1CaloRegion>"/>
   <class name="edm::Wrapper<std::vector<L1CaloEmCand>>" splitLevel="0"/>
   <class name="edm::Wrapper<std::vector<L1CaloRegion>>" splitLevel="0"/>
+
+  <class name="l1t::CICADABxCollection"/>
+  <class name="BXVector<float>"/>
```
Comment (fwyzard) [inline DataFormats/L1CaloTrigger/src/classes_def.xml:18]: isn't
```
<class name="BXVector<float>"/>
```
a duplicate of
```
<class name="l1t::CICADABxCollection"/>
```
?

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      cicadaBits[wordNum] =
+          (cicadaBitsPattern & raw_data) >>
+          28;  //The 28 shifts the extracted bits over to the start of the 32 bit result data for easier working with later
+    }
+    res_->push_back(
+        bx,
+        processBitsToScore(
+            cicadaBits));  //Now we insert CICADA into the proper BX, after a quick utility constructs a number from the 4 sets of bits.
+    ++processedBXs;  //index BXs
+  }
```
[the remark this answers is not in the document]
Comment (elfontan) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:66]: Hi @fwyzard,
documentation that I found recently is [here](https://github.com/cms-l1-globaltrigger/mp7_ugt_legacy/blob/master/doc/scales_inputs_2_ugt/pdf/scales_inputs_2_ugt.pdf) (see in particular Table 16 and 36).

Maybe something useful can be found also [here](https://github.com/cms-l1-globaltrigger/mp7_ugt_legacy/blob/master/doc/cicada_info/CICADA_Bit_Reference.pdf) and [here](https://github.com/cms-l1-globaltrigger/mp7_ugt_legacy/blob/master/doc/cicada_info/CICADA_Report.pdf).

I let @aloeliger and CICADA/Calo experts to comment more in detail about the past discussions with the GT Firmware team.

Cheers,
--Elisa

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      //Frames 4&5 are unused (by CICADA), they are reserved.
+      uint32_t raw_data = block.payload().at(wordLocation);
+      cicadaBits[wordNum] =
+          (cicadaBitsPattern & raw_data) >>
+          28;  //The 28 shifts the extracted bits over to the start of the 32 bit result data for easier working with later
+    }
+    res_->push_back(
+        bx,
+        processBitsToScore(
+            cicadaBits));  //Now we insert CICADA into the proper BX, after a quick utility constructs a number from the 4 sets of bits.
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:64]: This looks way overly complicated, I think it can all be simplified to
```suggestion
    // find the location of the first CICADA word
    unsigned short baseLocation = processedBXs * nFramesPerEvent;
    // read the 4 topmost bits from each CICADA word, and pack them into an integer
    uint32_t word = (cicadaBitsPattern & block.payload().at(baseLocation)) >> 16 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 1)) >> 20 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 2)) >> 24 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 3)) >> 28;
    // the CICADA score needs to be scaled by 8 bits, or 256
    float score = static_cast<float>(word) / 256.f;
    // insert the CICADA score into the proper BX
    res_->push_back( bx, score);
```

Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:64]: and if `block.payload()` is an `std::vector<uint32_t>`, it can be further simplified changing
```c++
    uint32_t word = (cicadaBitsPattern & block.payload().at(baseLocation)) >> 16 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 1)) >> 20 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 2)) >> 24 |
                    (cicadaBitsPattern & block.payload().at(baseLocation + 3)) >> 28;
```
to
```c++
    uint32_t* base = block.payload().data() + baseLocation;
    uint32_t word = (cicadaBitsPattern & base[0]) >> 16 |
                    (cicadaBitsPattern & base[1]) >> 20 |
                    (cicadaBitsPattern & base[2]) >> 24 |
                    (cicadaBitsPattern & base[3]) >> 28;
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+#include "DataFormats/L1CaloTrigger/interface/CICADA.h"
+
+#include "CaloSummaryUnpacker.h"
+#include "GTSetup.h"
+
+#include <cmath>
+
+float l1t::stage2::CaloSummaryUnpacker::processBitsToScore(const unsigned int bitsArray[]) {
+  float constructedScore = 0.0;
+  //All bits have been shifted to just left of the decimal point
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: Actually, reading the formula below I think the bits have been shifted so that
  - the bits go from `0x0000` to `0xffff`
  - the score goes from `0` to `255.996...`

Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: yes, the maximum score of CICADA is 0xff.ff ~= 255.996

Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: OK, then I think my simplified approach [above](https://github.com/cms-sw/cmssw/pull/44222#pullrequestreview-2139049212) should be correct.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:66]: Thank you, it matches with what the code seems to be doing (unlike the comments)

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc (hunk 8 above)
Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: It just wouldn't surprise me if the unpacker formula was over complicated, I was just writing off the top of my head, but it also demonstrably works. If it is going to be a sticking point, I can rework it to be less complicated.

Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: Try the simplified version...

Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: The problem is not only that the code works, but also that somebody will have to maintain it over the next years.

Comment (fwyzard) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc:17]: By the way, there seems to be some duplication between this code and that in https://github.com/cms-sw/cmssw/pull/45294 .

Maybe some can be reused ?

**Comment by mmusich** [comment]
@aloeliger this PR now has become urgent because @cms-sw/orp-l2 have merged https://github.com/cms-sw/cmssw/pull/45037 without this one leading to IB compilation failures. Can you adress Andrea's comments in order to convege quickly on it?

**Comment by aloeliger** [comment]
@mmusich it is a priority this morning, but I am also in the middle of weighing in on tests at point 5, and this needs to be carefully revalidated. @smuzaffar's revert is probably needed.

File: DataFormats/L1CaloTrigger/src/classes_def.xml (hunk 5 above)
Comment (aloeliger) [reply DataFormats/L1CaloTrigger/src/classes_def.xml]: Just a misunderstanding, this has been cleaned up.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc (hunk 7 above)
Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc]: This has been implemented, thanks for the suggestion.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc (hunk 8 above)
Comment (aloeliger) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloSummaryUnpacker.cc]: The simplification has been implemented, but there is a broader discussion to be had now about what happens to https://github.com/cms-sw/cmssw/pull/45294

**Comment by mmusich** [comment]
@aloeliger will you add here the revert of the revert (i.e. the revert of https://github.com/cms-sw/cmssw/pull/45308) directly here, so that we minimize issues in merging?

**Comment by aloeliger** [comment]
@mmusich @fwyzard I have implemented @fwyzard 's suggestions for simplifications to the unpacker logic (thanks for the suggestions!), and re-validated that it produces expected results. I will hold off on re-testing this until you're both satisfied with it's current form.

Now the question is about https://github.com/cms-sw/cmssw/pull/45037. This PR has the uGT unpacker (FED 1404 (production) & 1405 (test)) for CICADA, and the uGT emulator, which should be (to my understanding) the things that are necessary from the HLT side to connect CICADA to uGT. PR https://github.com/cms-sw/cmssw/pull/45037 has an unpacker for the Calo crate (FED 1356), which is required in the long term, but not required for running/connecting at least so far as HLT is concerned.

At some point that PR will need to be re-implemented/re-reverted/un-reverted whatever you want to call it. I can attempt to rebase this PR on top of master, include it again, and add the simplifications again, or I can package these separately. The question is just how quickly HLT needs this PR for a menu release. 

And speaking of menu releases/CMSSW_14_0, the other question is how you would like to handle the backports  https://github.com/cms-sw/cmssw/pull/45293 & https://github.com/cms-sw/cmssw/pull/45294.

**Comment by mmusich** [comment]
> At some point that PR will need to be re-implemented/re-reverted/un-reverted whatever you want to call it. I can attempt to rebase this PR on top of master, include it again, and add the simplifications again, or I can package these separately. 

see https://github.com/cms-sw/cmssw/pull/44222#issuecomment-2191279724

> The question is just how quickly HLT needs this PR for a menu release.

just to be clear, HLT *per se* doesn't need anything. The request comes from the CICADA / L1T developers, so I guess the answer is "as quickly you want it included".

> And speaking of menu releases/CMSSW_14_0, the other question is how you would like to handle the backports https://github.com/cms-sw/cmssw/pull/45293 & https://github.com/cms-sw/cmssw/pull/45294.

I'd lump them together in one of the two and close the other.

**Comment by aloeliger** [comment]
> see [#44222 (comment)](https://github.com/cms-sw/cmssw/pull/44222#issuecomment-2191279724)

Apologies, missed that. I will include everything here

> > The question is just how quickly HLT needs this PR for a menu release.
> 
> just to be clear, HLT _per se_ doesn't need anything. The request comes from the CICADA / L1T developers, so I guess the answer is "as quickly you want it included".
> 
Well, then speaking as the CICADA/L1T developer, let me rephrase that question. CICADA is angling to go into the next L1T menu (completely prescaled to 0 while remaining tests are finished), and my understanding is that HLT will need a release for that menu that includes this unpacker/emulator combo. I was at USCMS last week, and couldn't attend the ORP meeting, so I don't know when the next CMSSW_14_0 is expected, but I don't see an open build request. Does HLT have a planned release already for the next (august?) menu? Is getting this done today or tomorrow going to be too late for that release?

> > And speaking of menu releases/CMSSW_14_0, the other question is how you would like to handle the backports #45293 & #45294.
> 
> I'd lump them together in one of the two and close the other.

Done deal, I'll rework the backports.

**Comment by mmusich** [comment]
>  and my understanding is that HLT will need a release for that menu that includes this unpacker/emulator combo.

correct.

> so I don't know when the next CMSSW_14_0 is expected, but I don't see an open build request.

my understanding is that we're waiting for the L1T PRs (including this one) to coalesce before cutting a new release.

> Does HLT have a planned release already for the next (august?) menu? Is getting this done today or tomorrow going to be too late for that release?

it's not too late. The tentative deadline for HLT JIRA tickets is next Wednesday. I guess a ticket for CICADA can be placed ahead of the deadline even if the release is not physically available. We'll start integration when we have the (validated) ingredients at hand.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by aloeliger** [comment]
Okay, the recent update should simplify all unpacker logic, reverts the calo crate unpacker PR revert (adding it back), and should package all this together. I will try to mirror this to the extent I can with the open backports. @mmusich thank you for triggering the tests. I envy you your reflexes, how did you manage to hit the test on that immediately?

**Comment by mmusich** [comment]
> how did you manage to hit the test on that immediately?

luck, I had the tab open and noticed the force-push.

File: DataFormats/L1CaloTrigger/interface/CICADA.h [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#ifndef DataFormats_L1Trigger_CICADA_h
+#define DataFormats_L1Trigger_CICADA_h
+
+#include "DataFormats/L1Trigger/interface/BXVector.h"
+
+namespace l1t {
+  typedef BXVector<float> CICADABxCollection;
```
Comment (missirol) [inline DataFormats/L1CaloTrigger/interface/CICADA.h:7]: Coming back to this question https://github.com/cms-sw/cmssw/pull/44222#discussion_r1514854288, maybe from a different angle, I wonder if `float` is a good long-term choice, as opposed to a dedicated `l1t:CICADA` class (however simple it may currently be).

Some points.
 - Today, `l1t::CICADA` would just be something like (in short) `class CICADA { float score; };`. Is there a possibility that in future one could use the available bits to extract more than a score (e.g. a score with reduced precision plus some different info from other bits) ? If so, a dedicated data format would be more future-proof, and it would minimise changes downstream in the future.
 - On the topic of downstream changes (and minimising them in the future), I was asked by Efe if the 'CICADA object' should be propagated to the machinery that HLT uses for "trigger objects" (see for example the discussion in https://github.com/cms-sw/cmssw/pull/45295). I don't think this is strictly necessary, but if it will/should be done, having a dedicated class `l1t:CICADA` would be better, imho. @cms-sw/hlt-l2 can correct me.
 - In https://github.com/cms-sw/cmssw/pull/45227, there is a different-but-similar use case, and in that case a dedicated class is being introduced (`AXOL1TLScore`).

I do not know if this (`float` vs `l1t:CICADA`) needs to be addressed in this PR or not, but it's worth bringing up imho.

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
That's a phase 2 issue, which should be impossible for this to affect

**Comment by aloeliger** [comment]
+l1

**Comment by aloeliger** [comment]
Wrong PR. But whatever.

**Comment by iarspider** [comment]
> TC X1 = 10295.1 out of the seeding histogram bounds 0.076 - 0.58

Known issue: https://github.com/cms-sw/cmssw/issues/41451 .

File: DataFormats/L1CaloTrigger/interface/CICADA.h (hunk 9 above)
Comment (Martin-Grunewald) [reply DataFormats/L1CaloTrigger/interface/CICADA.h:7]: So this part of a dedicated data type is descoped?

Comment (aloeliger) [reply DataFormats/L1CaloTrigger/interface/CICADA.h:7]: That would be my preference.

Comment (Martin-Grunewald) [reply DataFormats/L1CaloTrigger/interface/CICADA.h:7]: OK!

**Comment by Martin-Grunewald** [comment]
+1

File: DataFormats/L1CaloTrigger/interface/CICADA.h (hunk 9 above)
Comment (aloeliger) [reply DataFormats/L1CaloTrigger/interface/CICADA.h:7]: Okay, just had a chat with @eyigitba about the technical reasons for doing this. Strictly, I think it is agreed it is not needed for this PR, but may be more integral for "sophisticated" use at HLT (treating the CICADA score as an object in case we want to do that) than I was giving it credit for. I think there is also desire to centralize with AXO some of these ideas. I can take a look at this with a bit more priority in that case.

**Comment by antoniovilela** [comment]
@aloeliger @cms-sw/l1-l2 
Just confirming that this includes the re-revert of https://github.com/cms-sw/cmssw/pull/45037?

**Comment by aloeliger** [comment]
> @aloeliger @cms-sw/l1-l2 
> Just confirming that this includes the re-revert of https://github.com/cms-sw/cmssw/pull/45037?

@antoniovilela That is correct. bcb0dabf2c9753a2119ad06b6d004ed463a4d404 should revert #45037 (the original revert)

**Comment by antoniovilela** [comment]
> > @aloeliger @cms-sw/l1-l2 
> > Just confirming that this includes the re-revert of https://github.com/cms-sw/cmssw/pull/45037?
> 
> @antoniovilela That is correct. bcb0dabf2c9753a2119ad06b6d004ed463a4d404 should rerevert #45037 (I.e. it is included again)

Thanks.

File: L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc [hunk 10]
```diff
@@ -51,6 +51,8 @@ void L1TGlobalProducer::fillDescriptions(edm::ConfigurationDescriptions& descrip
       ->setComment("InputTag for Calo Trigger EtSum (required parameter:  default value is invalid)");
   desc.add<edm::InputTag>("EtSumZdcInputTag", edm::InputTag(""))
       ->setComment("InputTag for ZDC EtSums Plus and Minus (required parameter:  default value is invalid)");
+  desc.add<edm::InputTag>("CICADAInputTag", edm::InputTag(""))
+      ->setComment("InputTag for CICADA Anomaly Detection (required parameter: default value is invalid)");
```
Comment (Martin-Grunewald) [inline L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:55]: So now that this PR and its backport are in 14_1 / 14_0, what value for this new parameter should be used at HLT level running the L1TGlobalProducer? ```hltGtStage2ObjectMap = cms.EDProducer( "L1TGlobalProducer",...```

Comment (aloeliger) [reply L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:55]: @Martin-Grunewald Do you want emulator or DAQ values?

Comment (Martin-Grunewald) [reply L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:55]: So in HLT where we run ```hltGtStage2Digis = cms.EDProducer( "L1TRawToDigi",...```
it would be ```hltGtStage2ObjectMap.CICADAInputTag = cms.InputTag("hltGtStage2Digis","CICADAScore"),```, correct?

Comment (mmusich) [reply L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:55]: Can this discussion be documented in a [CMSHLT](https://its.cern.ch/jira/projects/CMSHLT/issues) Jira ticket please @aloeliger  ?

Comment (Martin-Grunewald) [reply L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:55]: https://its.cern.ch/jira/browse/CMSHLT-3280
